engine/src/flutter/shell/platform/embedder/embedder.cc

```diff
@@ -2355,6 +2349,21 @@ FlutterEngineResult FlutterEngineInitialize(size_t version,
                               "Task runner configuration was invalid.");
   }
 
+  // Embedder supplied UI task runner runner does not have a message loop.
+  bool has_ui_thread_message_loop =
+      task_runners.GetUITaskRunner()->GetTaskQueueId().is_valid();
+  settings.task_observer_add = [has_ui_thread_message_loop](
+                                   intptr_t key, const fml::closure& callback) {
+    if (has_ui_thread_message_loop) {
+      fml::MessageLoop::GetCurrent().AddTaskObserver(key, callback);
+    }
+  };
+  settings.task_observer_remove = [has_ui_thread_message_loop](intptr_t key) {
+    if (has_ui_thread_message_loop) {
+      fml::MessageLoop::GetCurrent().RemoveTaskObserver(key);
+    }
+  };
```

Comment:
Just for my own curiosity, what are the task observers used for? I don't understand the implications of not registering these observers

Reply:
Ah from the comment in `EmbedderEngine::RunTask`, it sounds like the observers are used to flush microtasks? If so, what do you think adding a quick comment here to explain that this is used to flush microtasks if there's a separate UI thread, otherwise, microtasks are flushed by `EmbedderEngine::RunTask`?